### PR TITLE
graphql-schema-diff: replace async-graphql-parser with cynic-parser

### PR DESCRIPTION
**Pull request by tomhoule**
The main reason is that we are dealing with large GraphQL schemas where async-graphql-parser blows up in size and memory usage in production. cynic-parser doesn't.

The other main reason is that cynic-parser has a much nicer API and sounder foundations to build on.

closes GB-6465

### CODE REVIEW

File: engine/crates/graphql-schema-diff/src/state.rs [hunk 1]
```diff
@@ -266,3 +262,56 @@ fn push_schema_definition_changes(
...
+                    let Some((_, b_value)) = b.iter().find(|(b_name, _)| b_name == name) else {
+                        return false;
+                    };
+
+                    value_eq(value, b_value)
+                })
+        }
+        _ => false,
+    }
+}
```
Comment (tomhoule) [inline engine/crates/graphql-schema-diff/src/state.rs:317]: Should these last two be upstreamed in some way?

File: engine/crates/graphql-schema-diff/tests/diff/field_wrapping_type_changes.graphql [hunk 2]
```diff
@@ -0,0 +1,13 @@
+scalar Date
```
Comment (tomhoule) [inline engine/crates/graphql-schema-diff/tests/diff/field_wrapping_type_changes.graphql:1]: This wasn't tested directly in this crate, but in operation-checks. Added this test for tighter feedback loop when debugging.

File: engine/crates/graphql-schema-diff/src/state.rs (hunk 1 above)
Comment (obmarg) [reply engine/crates/graphql-schema-diff/src/state.rs:306]: Yes, these seem like they'd be useful.

File: engine/crates/graphql-schema-diff/src/state.rs [hunk 3]
```diff
@@ -266,3 +262,45 @@ fn push_schema_definition_changes(
...
+        (ast::Value::BlockString(a), ast::Value::BlockString(b)) => a == b,
+        (ast::Value::Boolean(a), ast::Value::Boolean(b)) => a == b,
+        (ast::Value::Null, ast::Value::Null) => true,
+        (ast::Value::Enum(a), ast::Value::Enum(b)) => a == b,
+        (ast::Value::List(a), ast::Value::List(b)) => {
+            a.len() == b.len() && a.iter().zip(b.iter()).all(|(a, b)| value_eq(a, b))
+        }
+        (ast::Value::Object(a), ast::Value::Object(b)) => {
+            a.len() == b.len()
+                && a.iter().all(|(name, value)| {
```
Comment (tomhoule) [inline engine/crates/graphql-schema-diff/src/state.rs:296]: this is wrong: b could have keys a doesn't have and this would still return true

Comment (obmarg) [reply engine/crates/graphql-schema-diff/src/state.rs:296]: Doesn't the equal length prevent that?

This looks fairly similar to the HashMap impl: https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#1289

Comment (tomhoule) [reply engine/crates/graphql-schema-diff/src/state.rs:296]: ah true, should have self-reviewed more carefully

**Review by obmarg — APPROVED** [review]

**Pull request merged**